# Analyzing borrowers’ risk of defaulting

<h2>Project Description


The project is to prepare a report for a bank’s loan division. The aim is to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

The report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

Description of the data
- children: the number of children in the family
- days_employed: how long the customer has been working
- dob_years: the customer’s age
- education: the customer’s education level
- education_id: identifier for the customer’s education
- family_status: the customer’s marital status
- family_status_id: identifier for the customer’s marital status
- gender: the customer’s gender
- income_type: the customer’s income type
- debt: whether the customer has ever defaulted on a loan
- total_income: monthly income
- purpose: reason for taking out a loan

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Analyzing-borrowers’-risk-of-defaulting" data-toc-modified-id="Analyzing-borrowers’-risk-of-defaulting-1">Analyzing borrowers’ risk of defaulting</a></span><ul class="toc-item"><li><span><a href="#Project-Description" data-toc-modified-id="Project-Description-1.1">Project Description</a></span></li><li><span><a href="#Open-the-data-file-and-have-a-look-at-the-general-information.-" data-toc-modified-id="Open-the-data-file-and-have-a-look-at-the-general-information.--1.2">Open the data file and have a look at the general information. </a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#-Conclusion" data-toc-modified-id="-Conclusion-1.2.0.1"> Conclusion</a></span></li></ul></li></ul></li><li><span><a href="#Preprocess-the-data:" data-toc-modified-id="Preprocess-the-data:-1.3">Preprocess the data:</a></span><ul class="toc-item"><li><span><a href="#Identify-and-fill-in-missing-values" data-toc-modified-id="Identify-and-fill-in-missing-values-1.3.1">Identify and fill in missing values</a></span><ul class="toc-item"><li><span><a href="#-Conclusion" data-toc-modified-id="-Conclusion-1.3.1.1"> Conclusion</a></span></li></ul></li><li><span><a href="#Replace-the-real-number-data-type-with-the-integer-type" data-toc-modified-id="Replace-the-real-number-data-type-with-the-integer-type-1.3.2">Replace the real number data type with the integer type</a></span><ul class="toc-item"><li><span><a href="#-Conclusion" data-toc-modified-id="-Conclusion-1.3.2.1"> Conclusion</a></span></li></ul></li><li><span><a href="#Delete-duplicate-data-" data-toc-modified-id="Delete-duplicate-data--1.3.3">Delete duplicate data </a></span><ul class="toc-item"><li><span><a href="#-Conclusion" data-toc-modified-id="-Conclusion-1.3.3.1"> Conclusion</a></span></li></ul></li><li><span><a href="#-Categorizing-Data" data-toc-modified-id="-Categorizing-Data-1.3.4"> Categorizing Data</a></span><ul class="toc-item"><li><span><a href="#-Conclusion" data-toc-modified-id="-Conclusion-1.3.4.1"> Conclusion</a></span></li></ul></li></ul></li><li><span><a href="#Loan-Repayment-Factors" data-toc-modified-id="Loan-Repayment-Factors-1.4">Loan Repayment Factors</a></span><ul class="toc-item"><li><span><a href="#Is-there-a-relation-between-having-kids-and-repaying-a-loan-on-time?" data-toc-modified-id="Is-there-a-relation-between-having-kids-and-repaying-a-loan-on-time?-1.4.1">Is there a relation between having kids and repaying a loan on time?</a></span><ul class="toc-item"><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.4.1.1">Conclusion</a></span></li></ul></li><li><span><a href="#Is-there-a-relation-between-marital-status-and-repaying-a-loan-on-time?" data-toc-modified-id="Is-there-a-relation-between-marital-status-and-repaying-a-loan-on-time?-1.4.2">Is there a relation between marital status and repaying a loan on time?</a></span><ul class="toc-item"><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.4.2.1">Conclusion</a></span></li></ul></li><li><span><a href="#-Is-there-a-relation-between-income-level-and-repaying-a-loan-on-time?" data-toc-modified-id="-Is-there-a-relation-between-income-level-and-repaying-a-loan-on-time?-1.4.3"> Is there a relation between income level and repaying a loan on time?</a></span><ul class="toc-item"><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.4.3.1">Conclusion</a></span></li></ul></li><li><span><a href="#How-do-different-loan-purposes-affect-on-time-repayment-of-the-loan?" data-toc-modified-id="How-do-different-loan-purposes-affect-on-time-repayment-of-the-loan?-1.4.4">How do different loan purposes affect on-time repayment of the loan?</a></span><ul class="toc-item"><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.4.4.1">Conclusion</a></span></li></ul></li></ul></li><li><span><a href="#General-conclusion" data-toc-modified-id="General-conclusion-1.5">General conclusion</a></span></li></ul></li></ul></div>

<h2>Open the data file and have a look at the general information. 

In [1]:
import pandas as pd

credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv')
print(credit_scoring.head(5)) 

#checking for the range  and kind of values
print(credit_scoring['children'].value_counts())
print(credit_scoring['dob_years'].value_counts())
print(credit_scoring['education'].value_counts())
print(credit_scoring['family_status'].value_counts())
print(credit_scoring['gender'].value_counts())
print(credit_scoring['income_type'].value_counts())
print(credit_scoring['purpose'].value_counts())

credit_scoring.info() #for checking the data type of the columns in the dataframe



   children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house 

<h4> Conclusion

- `days_employed` should have datatype as int. 
- `days_employed` and `total_income` have missing values, hence 19351 entries as compared to the total of 21525 entries 

- noticed `days_employed` value with days as -8139 kind of values, I think that these are the account holders who have been employed either before taking the loan or opening the bank account. 
- There are `days_employed` value which are more than 300000, that is around 800 years, so I assume it is incorrect data.

<h2>Preprocess the data:

<h3>Identify and fill in missing values

In [2]:
#checking for missing values
print(credit_scoring.isnull().sum())

#checking for days_employed values that are more than 0
days = credit_scoring[credit_scoring['days_employed']>=0]
print('\nNumber of days entries more than or equal to zero:',days['days_employed'].value_counts().sum())

#checking for total_income values 
zeroornegative_income = credit_scoring[credit_scoring['total_income'] <= 0]
print('\nNumber of total_income entries less than or equal to zero:',zeroornegative_income['total_income'].value_counts().sum())
income = credit_scoring[credit_scoring['total_income'] > 0]
print('\nNumber of total_income entries more than zero:',income['total_income'].value_counts().sum())

#calculating the median for dob_year column
age_values = credit_scoring[credit_scoring['dob_years']>0]
age_median = age_values['dob_years'].median()

#calculating the median of the days_employed column entries
days_values = credit_scoring[credit_scoring['days_employed']<0]
days_median = days_values['days_employed'].median()

#calculating the median of the total_income column entries
income_values = credit_scoring[credit_scoring['total_income']>0]
income_median = income_values['total_income'].median()

#Replacing the 0 values with age_median calculated above 
credit_scoring.loc[(credit_scoring['dob_years'] == 0), 'dob_years'] = age_median

#Replacing the missing values and the unrealistic values with days_median calculated above by using the fillna() method
credit_scoring['days_employed'] = credit_scoring['days_employed'].fillna(value = days_median)
credit_scoring.loc[(credit_scoring['days_employed']>0), 'days_employed'] = days_median

#Replacing the missing values and the unrealistic values with days_median calculated above by using the fillna() method
credit_scoring['total_income'] = credit_scoring['total_income'].fillna(value = income_median)


#Changing the negative value of children to a positive value using abs() function
credit_scoring['children'] = abs(credit_scoring['children'])


#Changing the negative value of days_employed to a positive value using abs() function
credit_scoring['days_employed'] = abs(credit_scoring['days_employed'])

print('\n')
print(credit_scoring.head())
print('\n')
print(credit_scoring.isnull().sum())


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Number of days entries more than or equal to zero: 3445

Number of total_income entries less than or equal to zero: 0

Number of total_income entries more than zero: 19351


   children  days_employed  dob_years            education  education_id  \
0         1    8437.673028       42.0    bachelor's degree             0   
1         1    4024.803754       36.0  secondary education             1   
2         0    5623.422610       33.0  Secondary Education             1   
3         3    4124.747207       32.0  secondary education             1   
4         0    1630.019381       53.0  secondary education             1   

       family_status  family_status_id gender income_ty

<h4> Conclusion

- Calculated `days_emloyed` mean by working on the entries that negative values based on my first conclusion
- Calculated `total_income` median by working on the entries where the debt was more than zero
- `days_employed` and `total_income` initially had 2174 missing values, replaced the missing values with median of that column
- Also replaced 101 - 0 age entries with the median of `age` column

<h3>Replace the real number data type with the integer type

In [3]:
#changing days_employed and dob_years datatype to int

credit_scoring['days_employed'] = credit_scoring['days_employed'].astype("int")
credit_scoring['dob_years'] = credit_scoring['dob_years'].astype("int")
credit_scoring.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


<h4> Conclusion

changing `days_employed` and `dob_years` datatype to int by using astype method

<h3>Delete duplicate data 

In [4]:
#changing the values to lower case for consistency
credit_scoring['education'] = credit_scoring['education'].str.lower()
credit_scoring['family_status'] = credit_scoring['family_status'].str.lower()
credit_scoring['gender'] = credit_scoring['gender'].str.lower()
credit_scoring['income_type'] = credit_scoring['income_type'].str.lower()
credit_scoring['purpose'] = credit_scoring['purpose'].str.lower()


#removing duplicates
credit_scoring = credit_scoring.drop_duplicates(keep = 'last').reset_index(drop=True)
credit_scoring.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 12 columns):
children            21454 non-null int64
days_employed       21454 non-null int64
dob_years           21454 non-null int64
education           21454 non-null object
education_id        21454 non-null int64
family_status       21454 non-null object
family_status_id    21454 non-null int64
gender              21454 non-null object
income_type         21454 non-null object
debt                21454 non-null int64
total_income        21454 non-null float64
purpose             21454 non-null object
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


<h4> Conclusion

Converted all string datatype column into lowercase and then dropped duplicates

<h3> Categorizing Data

In [5]:
#categorizing the total_income into different income_groups

income_median = credit_scoring['total_income'].median()
income_min = credit_scoring['total_income'].min()
income_max = credit_scoring['total_income'].max()

def income_groups(total_income):
    if income_min <= total_income <(income_median/2):
        return 'low income group'
    elif (income_median/2) <=  total_income <income_median:
        return 'middle income group'
    else:
        return 'high income group'

    
credit_scoring['income_groups'] = credit_scoring['total_income'].apply(income_groups)
print(credit_scoring['income_groups'].value_counts())



#categorizing purpose into much more compact categories
def purpose_category(purpose):
    if any(word in purpose for word in ['wedding','wed','marriage']):
        return 'wedding'
    elif any(word in purpose for word in ['property']):
        return 'property'
    elif any(word in purpose for word in ['house','housing']):
        return 'house'
    elif any(word in purpose for word in ['university','education']):
        return 'education'
    elif any(word in purpose for word in ['car']):
        return 'car'    
    else:
        return 'real estate'
    

credit_scoring['purpose_category'] = credit_scoring['purpose'].apply(purpose_category)
print(credit_scoring['purpose_category'].value_counts())
print(credit_scoring.head())
        

high income group      11779
middle income group     8014
low income group        1661
Name: income_groups, dtype: int64
real estate    4872
car            4306
house          3809
education      3605
property       2538
wedding        2324
Name: purpose_category, dtype: int64
   children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0           1630         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      f    employee     0     40620.102   
1            married                 0      f    employee     0     17932.802   
2            married  

<h4> Conclusion

Categorized `total_income` data by splitting it into 3 groups low, middle and high

Categorized purpose data by merging them into more compact 6 groups say buying house, renovating house, etc will all fall into house category

<h2>Loan Repayment Factors

<h3>Is there a relation between having kids and repaying a loan on time?

In [6]:
#calling pivot table method to find out the dependency between children and debt values. 
#Since sum is the summation of all 1 values we got all the defaulters from aggfunc sum and the total number of loan takers from count

children_pivot = credit_scoring.pivot_table(index = 'children', values = 'debt', aggfunc = ['count', 'sum'])
children_pivot['non-defaulters'] = children_pivot['count'] - children_pivot['sum']
children_pivot['ratio of defaulters'] = children_pivot['sum'] / children_pivot['count']

print(children_pivot.sort_values('ratio of defaulters', ascending = False))



          count   sum non-defaulters ratio of defaulters
           debt  debt                                   
children                                                
20           76     8             68            0.105263
4            41     4             37            0.097561
2          2052   194           1858            0.094542
1          4855   445           4410            0.091658
3           330    27            303            0.081818
0         14091  1063          13028            0.075438
5             9     0              9            0.000000


<h4>Conclusion

From the above result, I see that normally the more number of kids there are more defaulters. Assuming 0 in debt column represents non-defaulters and 1 being a defaulter.

<h3>Is there a relation between marital status and repaying a loan on time?

In [7]:
#calling pivot table method to find out the dependency between family_status and debt values. 
#Since sum is the summation of all 1 values we got all the defaulters from aggfunc sum and the total number of loan takers 
#from count


family_pivot = credit_scoring.pivot_table(index = 'family_status', values = 'debt', aggfunc = ['count', 'sum'])
family_pivot['non-defaulters'] = family_pivot['count'] - family_pivot['sum']
family_pivot['ratio of defaulters'] = family_pivot['sum'] / family_pivot['count']

print(family_pivot.sort_values('ratio of defaulters', ascending = False))

                   count  sum non-defaulters ratio of defaulters
                    debt debt                                   
family_status                                                   
unmarried           2810  274           2536            0.097509
civil partnership   4151  388           3763            0.093471
married            12339  931          11408            0.075452
divorced            1195   85           1110            0.071130
widow / widower      959   63            896            0.065693


<h4>Conclusion

From the above result - I conclude that the unmarried people are the highest number of defaulters and the lowest being widow/widower

<h3> Is there a relation between income level and repaying a loan on time?

In [8]:
#calling pivot table method to find out the dependency between income_groups and debt values. 
#Since sum is the summation of all 1 values we got all the defaulters from aggfunc sum and the total number of loan takers 
#from countprint(credit_scoring.groupby('income_groups')['debt'].value_counts())

income_groups_pivot = credit_scoring.pivot_table(index = 'income_groups', values = 'debt', aggfunc = ['count', 'sum'])
income_groups_pivot['non-defaulters'] = income_groups_pivot['count'] - income_groups_pivot['sum']
income_groups_pivot['ratio of defaulters'] = income_groups_pivot['sum'] / income_groups_pivot['count']

print(income_groups_pivot.sort_values('ratio of defaulters', ascending = False))

                     count  sum non-defaulters ratio of defaulters
                      debt debt                                   
income_groups                                                     
middle income group   8014  689           7325            0.085975
high income group    11779  937          10842            0.079548
low income group      1661  115           1546            0.069235


<h4>Conclusion

Middle income groups are highest defaulters, whereas high and low income group have a similar defaultment ratio

<h3>How do different loan purposes affect on-time repayment of the loan?

In [9]:
#calling pivot table method to find out the dependency between purpose_category and debt values. 
#Since sum is the summation of all 1 values we got all the defaulters from aggfunc sum and the total number of loan takers 
#from count

purpose_category_pivot = credit_scoring.pivot_table(index = 'purpose_category', values = 'debt', aggfunc = ['count', 'sum'])
purpose_category_pivot['non-defaulters'] = purpose_category_pivot['count'] - purpose_category_pivot['sum']
purpose_category_pivot['ratio of defaulters'] = purpose_category_pivot['sum'] / purpose_category_pivot['count']

print(purpose_category_pivot.sort_values('ratio of defaulters', ascending = False))


                 count  sum non-defaulters ratio of defaulters
                  debt debt                                   
purpose_category                                              
car               4306  403           3903            0.093590
education         3605  331           3274            0.091817
wedding           2324  186           2138            0.080034
real estate       4872  375           4497            0.076970
property          2538  190           2348            0.074862
house             3809  256           3553            0.067209


<h4>Conclusion

From the above result we see that the people who have taken out loans for cars are the highest number of defaulters and the lowest being for housing reason

<h2>General conclusion

- I think that the blank values in the `days_employed` and `total_income` are an outcome of some scheme or an offer provided to some company or selected group of individuals by the bank. As they are the same rows in which both the `days_employed` and `total_income` are blank at the same time. 
- Apart from that I also noticed that people who concentrate on basic neccessities say house and have 0-2 kids are the one who are less defaulters. 
- And also education is costly and there are not enough jobs in the market - that might also be a reason why there are loan defaulters.

Thank You!